### PR TITLE
add lecture 02 slides

### PR DESCRIPTION
**Pull request by andrei-shulaev**

### CODE REVIEW

**Comment by Cjkjvfnby** [comment]
Diff is not shown in GitHub, too many lines. See `To see what is in scope` section.

Please capitalize new lines.
- image
- image
- image


> Mutate string

Confusing title for creating a new instance of the string. 

> String formatting

Should we talk about formating arguments https://docs.python.org/3/library/string.html#formatspec?

> Nice alternative to map and filter

I'd reccoment skiping the part with map and filter.

Looks like github shows me the same image for all pictures.
image

> This behavior could be canged global and nonlocal operators

Missed preposition

**Review by cdeler — COMMENTED** [review]
As the diff is too large, I left the comments there:

1. I'd mention `NoneType` in this diagram "A mutable object can be changed after it's created, and an immutable object can't"
1. From my point of view it's better not to use `**kwargs` in the presentation (or to show what's that)
1. I'd format all the code in cells using `black` (using [blackcellmagic](https://github.com/csurfer/blackcellmagic) or [nbQA](https://github.com/nbQA-dev/nbQA))

**Comment by andrei-shulaev** [comment]
> Diff is not shown in GitHub, too many lines. See `To see what is in scope` section.

removed 
> Please capitalize new lines.

done
> Confusing title for creating a new instance of the string.

changed
> Should we talk about formating arguments https://docs.python.org/3/library/string.html#formatspec?
 
> > Nice alternative to map and filter
> 
> I'd reccoment skiping the part with map and filter.

Not agree. I'm sure students will have strong intention to overuse functional approach after FP lecture. They should know that using of list comprehensions is preferable. 

> Looks like github shows me the same image for all pictures.
> image

Some images are simply copy-pasted into the cell. So they have the same attachments name. It should work fine locally.

> Missed preposition

Changed

**Comment by andrei-shulaev** [comment]
>     1. I'd mention `NoneType` in this diagram "A mutable object can be changed after it's created, and an immutable object can't"
 
Will do it.

>     2. From my point of view it's better not to use `**kwargs` in the presentation (or to show what's that)

I do it when explaining packing/unpacking using double asterisk

>     3. I'd format all the code in cells using `black` (using [blackcellmagic](https://github.com/csurfer/blackcellmagic) or [nbQA](https://github.com/nbQA-dev/nbQA))

I would prefer to stay as is. Formatter fails on some cells and I have to make them shorter since I plan to show it in presentation mode.

**Comment by Cjkjvfnby** [comment]
>  I'm sure students will have strong intention to overuse functional approach after FP lecture. They should know that using of list comprehensions is preferable.

Then let's move it to the FP lecture.

**Comment by Cjkjvfnby** [comment]
> Some images are simply copy-pasted into the cell. So they have the same attachments name. It should work fine locally.

That the reason for the huge diff.
```
 "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtAAAAIGCAIA..............
```

**Comment by andrei-shulaev** [comment]
> Then let's move it to the FP lecture.

Removed. We have to make sure it would be mentioned during next lesson.

**Comment by andrei-shulaev** [comment]
> That the reason for the huge diff.

Now it is viewable.

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than ⌊ n/2 ⌋ times.
```
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw2.py]: May be rewrite it in python terms: 

```suggestion
The max element is the element that appears more than  n // 2 times.
```

**Review by cdeler — DISMISSED** [review]
Approved in advance

File: lecture_02_data_structures_functions/hw/hw1.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+from typing import List
+
+
+def get_longest_diverse_words(file_path: str) -> List[str, 10]:
```
Comment (Cjkjvfnby) [inline lecture_02_data_structures_functions/hw/hw1.py]: I haven't found any mention of such a size declaration in typing.List. 

```suggestion
def get_longest_diverse_words(file_path: str) -> List[str]:
```

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+The max element is the element that appears more than n // times.
+The min element is the element that appears fewer than other.
+
+You may assume that the array is non-empty and the max element always exist in the array.
+
+Example 1:
+
+Input: [3,2,3]
+Output: 3, 2
+Example 2:
```
Comment (Cjkjvfnby) [inline lecture_02_data_structures_functions/hw/hw2.py:14]:
```suggestion
Example 2:

````

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
```
Comment (Cjkjvfnby) [inline lecture_02_data_structures_functions/hw/hw2.py]:
```suggestion
Given an array of size n, find the most common and the least common elements.
```

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+The max element is the element that appears more than n // times.
```
Comment (Cjkjvfnby) [inline lecture_02_data_structures_functions/hw/hw2.py]: What is `times`?

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+Example 2:
+
+Input: [2,2,1,1,1,2,2]
+Output: 2, 1
+
+"""
+from typing import List, Tuple
+
+
+def major_and_minor_elem(inp: List) -> Tuple[int, 2]:
```
Comment (Cjkjvfnby) [inline lecture_02_data_structures_functions/hw/hw2.py]:
```suggestion
def major_and_minor_elem(inp: List[int]) -> Tuple[int, int]:
```
https://docs.python.org/3/library/typing.html#typing.Tuple

File: lecture_02_data_structures_functions/hw/hw2.py (hunk 5 above)
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw2.py]:
```suggestion
The most common is the element that appears more than n // 2 times.
```

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than n // times.
+The min element is the element that appears fewer than other.
```
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw2.py:4]:
```suggestion
The least common is the element that appears fewer than other.
```

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+You may assume that the array is non-empty and the max element always exist in the array.
```
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw2.py]:
```suggestion
The least common is the element that appears fewer than other.

You may assume that the array is non-empty and the most common element always exists in the array.
```

File: lecture_02_data_structures_functions/hw/hw4.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+"""
+Write a function that accept another function as an argument. Then it
```
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw4.py]:
```suggestion
Write a function that accepts another function as an argument. Then it
```

File: lecture_02_data_structures_functions/hw/hw2.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+"""
+Given an array of size n, find the most common and the least common elements.
+The most common element is the element that appears more than n // 2 times.
+The least element is the element that appears fewer than other.
```
Comment (cdeler) [inline lecture_02_data_structures_functions/hw/hw2.py]:
```suggestion
The least common element is the element that appears fewer than other.
```

**Review by cdeler — APPROVED** [review]

**Pull request merged**